### DISCUSSION TITLE
how to remove all children of a TabbedContent

### DISCUSSION THREAD

**Discussion by nnako** [Textualize/textual#5299] [Q&A] [answered]
In my application, I do the following:

- create a `TabbedContent`

this automatically seems to create a default `TabPane` and sets a CSS `id` attribute for it. Now, I wand to repeatedly remove existing `TabPane` elements and create new ones below the `TabbedContent` element. When using the `<TabbedContent>.clear_panes()` method, all the respective `TabPane` elements disappear, but when creating new `TabPane` elements by the `<TabbedContent>.add_pane()` function, the `id` attribute seems to get re-used. As by using `clear_panes()`, the old `TabPane` elements still seem to reside within memory, I get an error saying:

```
DuplicateIds: Tried to insert a widget with ID '--content-tab-tab-2', but a
widget ContentTab(id='--content-tab-tab-2') already exists with that ID in this
list of children. The children of a widget must have unique IDs.
```

There seems to be the `<TabbedContent>.remove_pane()` method for this, but it seems to require the `pane_id` of a specific `TabPane` element. I have not yet found a method which returns **all** of the `TabPane` elements below a `TabbedContent` element. The API documentation provides a method to get the "active" one, though.

Is it possible to remove (completely delete) all of a `TabbedContent`'s children at once? Or is there at least a method which returns a list of all children of a `TabbedContent` element, so that I would be able to remove them within a loop?

Thanks.

**Comment by TomJGooding** [comment]
Hopefully that `DuplicateIds` issue will be fixed soon - there's an existing issue here:

- #5215

**Comment by willmcgugan** [comment]
Are you awaiting `clear_panes` ?

**Reply by nnako** [reply]
nope. my code looks like this:

```python
# remove all existing TabPanes
_TabbedContent.clear_panes()

# add tab panes respective to current profiles
for _profile in profiles.keys():
    _TabPane = _TabbedContent.add_pane(
        TabPane(
            _profile,
            Static(
                profiles[_profile]["description"],
                classes="profile",
                ),
            name=_profile,
            )
        )
```

**Reply by willmcgugan** [reply, marked as answer]
Things that involve modifying the DOM don't happen immediately. If you want to ensure that `clear_panes` is finished, you will need to make your method a coroutine (`async def my_method(self):`), and `await` the result.

i.e. If you do this, it will ensure all panes are cleared before you add new ones.

```python
await _TabbedContent.clear_panes()
```

If you do that, I don't think the issue @TomJGooding linked to will apply.

**Reply by nnako** [reply]
Thanks. This worked.
